app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,26 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      flash[:notice] = "devise.omniauth_callbacks.success", :kind => "Google"
+      sign_in_and_redirect @user, :event => :authentication
+    else
+      session["devise.google_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def facebook
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      flash[:notice] = "devise.omniauth_callbacks.success", :kind => "Facebook"
+      sign_in_and_redirect @user, :event => :authentication
+    else
+      session["devise.facebook_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
```

Comment:
This post suggests that there are certain weird failure scenarios where Devise will try to call this method, so it may be safer to include it: http://stackoverflow.com/questions/10173833/devise-omniauth-failure-how-to-debug-it